scipy/special/basic.py

```diff
@@ -2259,15 +2288,41 @@ def factorial(n, exact=False):
     >>> arr = np.array([3, 4, 5])
     >>> factorial(arr, exact=False)
     array([   6.,   24.,  120.])
+    >>> factorial(arr, exact=True)
+    array([  6,  24, 120])
     >>> factorial(5, exact=True)
     120L
 
     """
     if exact:
-        return math.factorial(n)
+        if np.ndim(n) == 0:
+            return math.factorial(n)
+        else:
+            n = asarray(n)
+            un = np.unique(n)
+
+            if any(abs(un.astype(int)) != un):
+                raise ValueError("Input must be non-negative integers for "
+                                 "'exact' mode.")
```

Comment:
That the return values changed for negative inputs is an oversight rather than some planned thing.  There's no reason not to return 0 here if that's the backward compatible result.